1_LongestSubstring.cpp

```diff
@@ -0,0 +1,125 @@
+/*
+ Microsoft Interview Coaching Pilot Program MICP 2018
+Author: Akshama
+Language: C++
+
+/**********WEEK ONE***********/
+/*
+Given a string, find the length of the longest substring without repeating characters.
+
+Examples:
+
+Given "abcabcbb", the answer is "abc", which the length is 3.
+
+Given "bbbbb", the answer is "b", with the length of 1.
+
+Given "pwwkew", the answer is "wke", with the length of 3. Note that the answer must be a substring, "pwke" is a subsequence and not a substring.
+
+
+//TEBOW IT
+
+/*
+1. T - TALK
+Input: A string which may contain alphabets, numbers or special characters.
+Output: An integer (length of the longest substring without repeating characters)
+
+Clarified:
+Handling Special Characters and Numbers - Handled in the same way as characters
+Case sensitivity - Yes
+Space separated strings - Considered Valid
+
+2. E - EXAMPLES
+Sample Input  ---- Output
+abcabcbb            3
+bbbb                1
+pwwkew              3
+Cbcdb               4
+*empty string*      0
+abc def             7
+
+
+3. B - BRUTE-FORCE
+
+Check for all substrings starting from each letter of the string and return the longest substring with no repeating characters.
+For n*(n+1)/2 substrings, linear scanning would be done for unique characters with a time complexity of O(n^3).
+4. O - OPTIMISE
+Create a sliding window type of structure that updates the start pointer on each repetition of a character. A substring which is already checked isn't searched again.
+Optimised time complexity is O(n).
+5. W - WALK THROUGH
+Explained as comments in the code
+6. I - IMPLEMENT
+*/
+#include<iostream>
+#include<string>
+#include<utility>
+#include<vector>
+
+using namespace std;
+
+int longestSubstring(string str)
+{
+
+    vector<int> visited(256,-1);
+    //length of the current substring without any repeating character
+    int curr=0;
+    //length of the longest substring without any repeating character
+    int longest=0;
+    for(int i=0;i<str.length();i++)
+    {
+
+        //if this character is not yet processed/ processed but not present in current substring which is started again
+        if(visited[str[i]]==-1 || visited[str[i]]<(i-curr) )
+        {
+            curr++;
+            //Update the visited array value with the current index of the character
+            visited[str[i]]=i;
+
+            if(curr>longest)
+                longest=curr;
+        }
+
+        else
+        {
+            //case of repetition of a character
+            //reset the length of current substring without any repeating character and update the visited array value with the current index of the character.
+            curr=i-visited[str[i]];
+            visited[str[i]]=i;
+        }
+
+    }
+
+    return longest;
+}
+
+//7. T - TEST
+
+bool testlongestSubstring()

```

Comment:
I think I have a test case that will fail given you implementation.

Can you try the following input string and let me know what happens, and fix it if there are any errors? :)
Input: "abcadef"
Expected output: 6

Reply:
Thanks a lot for the code review Omeed! :)

I've tried this input string and I'm getting the output as 6 which is expected. 
Here's how my code logic works for this string:
0 1 2 3 4 5 6 
a b c a d e f

Initially, 
curr = 0
longest = 0
visited[x] = -1  for all x from 0 to 256 (essentially, all the ASCII symbols)

**i=0**           
visited[a] = visited[97] = -1
Hence, entering the 'if' block.
curr = 1
visited[a] = 0
longest = 1

**i=1**           
visited[b] = visited[98] = -1
Hence, entering the 'if' block.
curr = 2
visited[b] = 1
longest = 2

**i=2**           
visited[c] = visited[99] = -1
Hence, entering the 'if' block.
curr = 3
visited[c] = 2
longest = 3

**i=3**           
if(visited[a] == -1 || visited[a] < ( i -curr ))
Now, visited[a] = 0 and (i-curr) = (3-3) =0
Hence, condition not satisfied; entering 'else' block.
curr = (i - visited[a]) = (3-0) = 3
visited[a] = i = 3

**i=4**           
visited[d] = visited[100] = -1
Hence, entering the 'if' block.
curr = 4
visited[d] = 4
longest = 4

**i=5**    
visited[e] = visited[101] = -1
Hence, entering the 'if' block.
curr = 5
visited[e] = 5
longest = 5

**i=6**    
visited[f] = visited[102] = -1
Hence, entering the 'if' block.
curr = 6
visited[f] = 6
longest = 6

**for loop ends, returns longest which is 6 in the end, as shown above.**